### DISCUSSION TITLE
How to iterate over a list of strings and execute a method for each string as long as an error occurs

### DISCUSSION THREAD

**Discussion by dirkil** [louthy/language-ext#1262] [Q&A] [answered]
Suppose you have a method `Fin<Unit> Calc(string input)` and a list of strings that you want to call `Calc `on. `Calc` can fail and when that happens you want to stop iterating and return the error. I know you can do it with a `foreach` loop like that:
```
public Fin<Unit> Process(List<string> inputValues)
{
        foreach (var input in inputValues)
        {
            var result = Calc(input);
            if (result.IsFail)
                return result;
        }

        return Unit.Default;
}
```

This approach is rather non-functional and not in the style of LanguageExt. Is there a more functional version that someone can recommend?

**Comment by olegdunkan** [comment]
`public Fin<IEnumerable<Unit>> Process(List<string> values) => values.Select(Calc).Traverse(x => x);`

**Comment by louthy** [comment]
```c#
public Fin<Unit> Process(IEnumerable<string> inputValues) =>
    inputValues.Sequence(Calc).Map(_ => unit);

public Fin<Unit> Process(IEnumerable<string> inputValues) =>
    inputValues.Map(Calc).Traverse(identity).Map(_ => unit);

```
To get back to a `Fin<Unit>` you can use both `Sequence` and `Traverse`, which are two functions have similar capabilities and work with nested monadic types (particularly useful with collection monads); so the answer from @olegdunkan is reasonable but doesn't get you back to a `Fin<Unit>`.

As you can see from  above, `Sequence` is a little more concise for your use-case.

It's always worth remembering that `Sequence` and `Traverse` allow you to 'flip' the nested monadic types; so in this case get the `Fin` inner monad to be the outer monad.  And to make the `Enumerable` outer monad to become the inner monad.  This is so useful as a technique that if you work with monads a lot then you'll find yourself reaching for `Sequence` and `Traverse` all the time.

There are literally 100s of hand written `Sequence` and `Traverse` extension methods that have been optimised to work with the monads in language-ext, so you'll find they're a pretty efficient way to handle situations like this.

There's also `SequenceParallel`, `SequenceSerial`, `TraverseParallel`,  and `TraverseSerial` for working with asynchronous monads - they give a level of control over the amount of parallel processing you'll allow.  Especially good for lazy sequences of asynchronous values, where you may want to only launch a fixed number of threads processing the items in the sequence.

**Reply by dirkil** [reply]
First of all, thank you guys for taking the time to answer my question. 

I was well aware of `Traverse ()` and `Sequence()` but this is not exactly solving the problem that I have. The requirement was that the iteration over the input values should stop as soon as the first call to `Calc()` returns an error. 

When you project it with `Select()` the calculation is executed for each and every item and only after that `Traverse()` stops on the first `Error`. Imagine `Calc()` to be an expensive calculation you wouldn't want it to be executed if not necessary.

I was wondering if there was something like a `Fin<Unit> Traverse(this IEnumerable <A> ma, Func<A, Fin<B>>)` which iterates over all elements, applies the function to it as long as the return value `Fin<B>.IsFail == false`. When a `Fail` occurs it returns that `Error` and if no `Error` occurs it returns `Unit`.

Something like that:
```
public static Fin<Unit> Traverse<A, B>(this IEnumerable<A> ma, Func<A, Fin<B>> f)
{
      foreach (A input in ma)
      {
        var x = f(input);
        if (x.IsFail)
          return x;
      }

      return Unit.Default;
}
```

I know that this is messing around with side effects and for that purpose there is probably something available in the `Eff` area. But to my shame, I must confess that I don't know much about this area of LanguageExt yet.

So, what are your thoughts on that?

**Reply by louthy** [reply]
> I was well aware of Traverse () and Sequence() but this is not exactly solving the problem that I have. The requirement was that the iteration over the input values should stop as soon as the first call to Calc() returns an error.

This is exactly what happens with lazy sequences.

**Reply by louthy** [reply]
If you look at [the implementation of `Traverse`](https://github.com/louthy/language-ext/blob/main/LanguageExt.Core/Transformer/Traverse/FinT.cs#L151) you can see it does almost exactly what you're asking for:

**Comment by dirkil** [comment]
Hm, I don't get it to work with `Traverse()`; maybe you can help me out with a sample that demonstrates it.

I get it to work with `Aggregate()` but I am not sure if that is as readable as it could be:
```
using LanguageExt;
using LanguageExt.Common;
using static LanguageExt.Prelude;

var werte = new List<string> { "adsghdf", "sadghdfjds", "fail", "872r87" };
var result = Process(werte);

Fin<Unit> Process(List<string> inputValues)
{
    return inputValues.Aggregate(
        seed: Fin<Unit>.Succ(unit),
        func: (acc, input) => acc.IsFail ? acc : Calc(input)
    );
}


Fin<Unit> Calc(string input)
{
    return input == "fail" ? Error.New("FEHLER") : unit;
}
```

In this example `Calc()` is only called for the first three input values exactly as I want it to have but I think it is not so obvious to the reader, at least you have to think about it for a minute. Can someone come up with a better solution that uses `Traverse()`?

**Reply by louthy** [reply, marked as answer]
Pretty sure all of this is covered with the previous replies.  Did you try them?

But using your example, with some added `Console.WriteLine` logging, you can see it implemented (as in the previous reply) using `Sequence` and `Traverse`.  `Sequence` being slightly more concise.

```c#
using LanguageExt;
using LanguageExt.Common;
using static LanguageExt.Prelude;

var werte = new List<string> { "adsghdf", "sadghdfjds", "fail", "872r87" };
var result1 = Process1(werte);
var result2 = Process2(werte);

Console.WriteLine(result1);
Console.WriteLine(result2);

Fin<Unit> Process1(List<string> inputValues) =>
    inputValues.Sequence(Calc).Map(_ => unit);

Fin<Unit> Process2(List<string> inputValues) =>
    inputValues.Map(Calc).Traverse(identity).Map(_ => unit);

Fin<Unit> Calc(string input)
{
    Console.WriteLine($"Processing item: {input}");
    return input == "fail" ? Error.New("FEHLER") : unit;
}
```

The above outputs:

```
Processing item: adsghdf
Processing item: sadghdfjds
Processing item: fail
Processing item: adsghdf
Processing item: sadghdfjds
Processing item: fail
Fail(FEHLER)
Fail(FEHLER)
```
Which shows that the processing stops upon a `fail` item.

And just to be even clearer.  Your version will keep going even after a failure and will need to continue to consume the entire list before quitting.  `Sequence` and `Traverse` will always stop as soon as there's a fail-state.

**Reply by dirkil** [reply]
Hi Paul,

I am sorry for having had a mind block. What I didn't realize was that `inputValues.Map(Calc)` does nothing until `Traverse()` is called on it and then `Traverse()` can obviously break on the first occurrence of a Fail. 

Fun fact:
When I checked that before in the debugger, it showed me four results but I didn't realize that the debugger — for showing the results — needs to enumerate the results right at this time and thereby executing `Calc()`.  (see picture)

![image](https://github.com/louthy/language-ext/assets/12883649/1b25acb2-945d-4ecb-9bdd-c64b7b297d41)

That was a very important lesson for me and I guess for others as well. Thank you very much for your patience.
